astropy/table/column.py

```diff
@@ -1030,10 +1030,26 @@ def insert(self, obj, values, axis=0):
             data = np.insert(self, obj, None, axis=axis)
             data[obj] = values
         else:
-            # Explicitly convert to dtype of this column.  Needed because numpy 1.7
-            # enforces safe casting by default, so .  This isn't the case for 1.6 or 1.8+.
-            values = np.asarray(values, dtype=self.dtype)
-            data = np.insert(self, obj, values, axis=axis)
+            new_dtype = self.dtype
+            print(self.dtype.kind)
+
+            if self.dtype.kind in ('U', 'S'):
+                import pdb; pdb.set_trace()
+                # Convert input ``value`` to the string dtype of this column and
+                # find the length of the longest string in the array.
+                values_str_len = np.char.str_len(values).max()
+
+                # Parse the array-protocol typestring (e.g. '|U15') of self.dtype which
+                # has the character repeat count on the right side.
+                self_str_len = dtype_bytes_or_chars(self.dtype)
+                if values_str_len > self_str_len:
+                    new_dtype = self.dtype.kind + str(values_str_len)
+
+        # Explicitly convert to dtype of this column.  Needed because numpy 1.7
+        # enforces safe casting by default, so .  This isn't the case for 1.6 or 1.8+.
+        values = np.asarray(values, dtype=new_dtype)
+        data = np.insert(self.astype(new_dtype), obj, values, axis=axis)
```

Comment:
Add a `copy=False` inside the `astype`, otherwise you make an extra copy also if the dtype is unchanged.

Reply:
Different implementation so this does not apply now.